mm2src/coins/coin_balance.rs

```diff
@@ -0,0 +1,452 @@
+    let coin = lp_coinfind_or_err(&ctx, &req.coin).await?;
+    match coin {
+        MmCoinEnum::UtxoCoin(utxo) => utxo.hd_account_balance_rpc(req.params).await,
+        MmCoinEnum::QtumCoin(qtum) => qtum.hd_account_balance_rpc(req.params).await,
+        _ => MmError::err(HDAccountBalanceRpcError::ExpectedHDWalletDerivationMethod {
+            coin: coin.ticker().to_owned(),
+        }),
+    }
+}
+
+pub async fn check_hd_account_balance(
+    ctx: MmArc,
+    req: CheckHDAccountBalanceRequest,
+) -> MmResult<CheckHDAccountBalanceResponse, HDAccountBalanceRpcError> {
+    let coin = lp_coinfind_or_err(&ctx, &req.coin).await?;
+    match coin {
+        MmCoinEnum::UtxoCoin(utxo) => utxo.check_hd_account_balance_rpc(req.params).await,
+        MmCoinEnum::QtumCoin(qtum) => qtum.check_hd_account_balance_rpc(req.params).await,
+        _ => MmError::err(HDAccountBalanceRpcError::ExpectedHDWalletDerivationMethod {
+            coin: coin.ticker().to_owned(),
+        }),
+    }
+}
+
+pub mod common_impl {
+    use super::*;
+    use crate::hd_wallet::HDAddress;
+    use common::calc_total_pages;
+
+    pub(crate) async fn enable_hd_wallet_balance<Coin, Address, HDWallet, HDAccount, HDAccountChecker>(
+        coin: &Coin,
+        hd_wallet: &HDWallet,
+    ) -> BalanceResult<HDWalletBalance>
+    where
+        Coin: HDWalletCoinOps<Address = Address, HDWallet = HDWallet, HDAccount = HDAccount>
+            + HDWalletBalanceOps<
+                Address = Address,
+                HDWallet = HDWallet,
+                HDAccount = HDAccount,
+                HDAddressChecker = HDAccountChecker,
+            > + Sync,
+    {
+        let mut accounts = coin.get_accounts_mut(hd_wallet).await;
+        let gap_limit = coin.gap_limit(hd_wallet);
+        let address_checker = coin.produce_hd_address_checker().await?;
+
+        let mut result = HDWalletBalance {
+            accounts: Vec::with_capacity(accounts.len()),
+        };
+
+        for (account_index, hd_account) in accounts.iter_mut() {
+            let addresses = coin
+                .check_hd_account_balance(hd_account, &address_checker, gap_limit)
+                .await?;
+
+            let total_balance = addresses.iter().fold(CoinBalance::default(), |total, addr_balance| {
+                total + addr_balance.balance.clone()
+            });
+            let account_balance = HDAccountBalance {
+                account_index: *account_index,
+                derivation_path: RpcDerivationPath(coin.account_derivation_path(hd_account)),
+                total_balance,
+                addresses,
+            };
+
+            result.accounts.push(account_balance);
+        }
+
+        Ok(result)
+    }
+
+    pub async fn hd_account_balance_rpc<Coin, Address, HDWallet, HDAccount>(
+        coin: &Coin,
+        params: HDAccountBalanceParams,
+    ) -> MmResult<HDAccountBalanceResponse, HDAccountBalanceRpcError>
+    where
```

Comment:
I think these trait bounds can be also simplified.

Reply:
Done